### PR TITLE
feat: try continuing partially downloaded files

### PR DESCRIPTION
**Pull request by de-sh**
Closes #

### Why?

Download behavior should support continuing downloads that were partially completed earlier.

### Trials Performed

Use [download_qa.sh](https://gist.github.com/de-sh/1bfa3c727c9dc9af77951e43d64e097b) to perform the network reset scenario required for circumstances where download must be restarted.

REFER
- [Logs of downloader running into failure and restarting](https://gist.github.com/de-sh/e8346d6d8aea55ed5e58f232d511f335#file-downloader-log-L49)
- Screenshot of md5sum run on file that was to be downloaded, compared against the file that ended up being downloaded:
Screenshot from 2023-09-13 15-00-54

### CODE REVIEW

File: uplink/src/collector/downloader.rs [hunk 1]
```diff
@@ -308,46 +300,46 @@ impl FileDownloader {
         Ok((file, file_path))
     }
 
+    // Retry mechanism tries atleast 3 times before returning a download error
+    async fn retry_thrice(&mut self, url: &str, mut download: DownloadedFile) -> Result<(), Error> {
```
Comment (tekjar) [inline uplink/src/collector/downloader.rs:304]: We don't need retry_thrice. We can indefinitely retry (probably with 1s poll) until timeout

Comment (de-sh) [reply uplink/src/collector/downloader.rs:304]: We can do that in a separate PR I feel like

File: uplink/src/collector/downloader.rs [hunk 2]
```diff
@@ -308,46 +300,45 @@ impl FileDownloader {
         Ok((file, file_path))
     }
 
+    // Retry mechanism tries atleast 3 times before returning a download error
```
Comment (tekjar) [inline uplink/src/collector/downloader.rs:303]: Did we move this function down? It's hard to compare this with the deleted `read_thrice` code above

File: uplink/src/collector/downloader.rs [hunk 3]
```diff
@@ -378,6 +369,52 @@ pub struct DownloadFile {
     pub download_path: Option<PathBuf>,
 }
 
+// A temporary structure to help us retry downloads
+// that failed after partial completion.
+struct RetryDownloadHandle {
```
Comment (tekjar) [inline uplink/src/collector/downloader.rs:374]: Why is this called `RetryDownloadHandle`? I don't see any retry code

Comment (de-sh) [reply uplink/src/collector/downloader.rs:374]: As the comment states, this is used to encapsulate details regarding the download that will be useful in tracking the download and retrying when it fails to complete

File: uplink/src/collector/downloader.rs (hunk 2 above)
Comment (de-sh) [reply uplink/src/collector/downloader.rs:303]: Sagar had implemented retries from when the file was being created, with the changes brought about in this PR, it is obvious that we have to retry only for network errors which only occur during the download and write stage, which is this was move had to be made.

Comment (tekjar) [reply uplink/src/collector/downloader.rs:303]: Didn't understand. How is this related to moving `retry_thrice` from line `187` to line `304`

File: uplink/src/collector/downloader.rs (hunk 3 above)
Comment (tekjar) [reply uplink/src/collector/downloader.rs:374]: You don't name structs and methods based on callers. You name them based on what they do

Comment (tekjar) [reply uplink/src/collector/downloader.rs:374]: Rename this to `DownloadState` and make this member of `FileDownloader`

File: uplink/src/collector/downloader.rs [hunk 4]
```diff
@@ -312,42 +324,17 @@ impl FileDownloader {
     async fn download(
         &mut self,
         resp: Response,
-        mut file: File,
-        content_length: usize,
+        download: &mut DownloadState,
```
Comment (tekjar) [inline uplink/src/collector/downloader.rs:327]: Make `DownloadState` a member of `FileDownloader`. This is a mutable method with side-effects any way. No need to pass an separate object as argument for a different state

Add clear method to `DownloadState` to reset on finishing download

**Pull request merged**